### PR TITLE
Python: insecure-protocol outdated

### PR DESCRIPTION
**Pull request by yoff**
The recommended usage of [the ssl module](https://docs.python.org/3/library/ssl.html?highlight=ssl#module-ssl) has changed, leading to false positives from `py/insecure-protocol`.
FP report here: https://github.com/github/codeql/issues/2554

### CODE REVIEW

**Comment by yoff** [comment]
One approach is to exclude the case `PROTOCOL_TLS` from this query and instead handle it separately by taint, being sanitized by restricting the version (either via `OP_NO_TLSv1` or `SSLContext.minimum_version`).
In any case, it does seem to require some version specificity.

**Comment by yoff** [comment]
I wonder what we should actually warn about here. The Python documentation itself seems slightly conflicted. For instance, reading [here](https://docs.python.org/3/library/ssl.html?highlight=ssl#protocol-versions), we see that SSL 2 and 3 are disabled by default and "considered insecure". But their example program also disables TLS 1 and 1.1 which both [have known vulnerabilities](https://www.packetlabs.net/tls-1-1-no-longer-secure/).
Even TLS 1.2 [may have some issues](https://access.redhat.com/articles/2112261), but seems to be recommended for use, and 1.3 seems to not be fully supported yet.

In short we could go three ways:
 - Consider just SSL insecure, then default constructors will create secure connections.
 - Consider SSL as well as TLS 1.0 and 1.1 insecure, then constructed contexts will need modification.
 - Consider all but TLS 1.3 insecure, then most programs will be in trouble.

While the third option may be the most future proof, it does not feel practical. Incidentally, the FP post originating this investigation is in-between the first two options, considering TLS 1.0 insecure but not 1.1. I guess the world is just moving quickly...

**Comment by RasmusWL** [comment]
> Consider SSL as well as TLS 1.0 and 1.1 insecure, then constructed contexts will need modification.

I think you're on the to right thing here. From my point of view, this should give an alert if you end up using any SSL version, or TLS < 1.2.

The `py/insecure-default-protocol` query probably also needs to be updated, to handle the fact that you can get a secure context by using (with Python3.6+)
```py
ctx = ssl.SSLContext()
ctx.options |= ssl.OP_NO_TLSv1
ctx.options |= ssl.OP_NO_TLSv1_1
```

but since many people are still using Python3.5, in an ideal world, I still think we should be alerting if someone does
```py
ctx = ssl.SSLContext()

# using `ctx`, which could be something like
ctx.load_default_certs()
http.client.HTTPSConnection(hostname, context=ctx)
```

Documentation also badly needs an update :grimacing: 

https://github.com/github/codeql/blob/1af503ab84f21713abaa60bae955392cfbbfbdad/python/ql/src/Security/CWE-327/InsecureDefaultProtocol.qhelp#L15-L17

https://github.com/github/codeql/blob/1af503ab84f21713abaa60bae955392cfbbfbdad/python/ql/src/Security/CWE-327/InsecureProtocol.qhelp#L15-L17

> Even TLS 1.2 may have some issues, but seems to be recommended for use, and 1.3 seems to not be fully supported yet.

From reading around the web, I agree that TLS 1.2 have some issues. I found https://www.cloudflare.com/learning/ssl/why-use-tls-1.3/ which highlights a few weaknesses of TLS 1.2.

I don't know enough about the subject to give a definite answer. My understand is that using TLS 1.2 _can_ be safe if you set up things correctly (disallow insecure ciphers), and _can_ be unsafe if you don't.

**Comment by yoff** [comment]
Thanks for connecting me to the `security-lab` slack channel. Consensus there seems to be that a general query should allow TLS 1.2 and above so that will be the aim of this PR.

**Comment by RasmusWL** [comment]
Found a FP by chance: https://lgtm.com/projects/g/home-assistant/core/?mode=tree&ruleFocus=1507248366243 -- notice that the second instance allows TLS 1.0 and 1.1, so is a TP.

The link to https://wiki.mozilla.org/Security/Server_Side_TLS, which seems to be a good reference that defines what should be considered safe and what shouldn't. Would be great to include that reference in our qhelp (and probably also just next to the logic in the query that decides whether something is acceptable or not).

**Comment by yoff** [comment]
Thanks, that is a good link.

**Comment by yoff** [comment]
Did the merge and pushed my wip query.

**Comment by yoff** [comment]
Closing for now as this is getting out-dated.
